espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/Http/Router.java

```diff
@@ -1,113 +1,192 @@
     }
 
     private void addRoute(Method method, String uri, RequestHandler handler) throws DuplicateRouteException {
+        addRoute(method, uri, handler, AppiumParams.class);
+    }
+
+    private void addRoute(Method method, String uri, RequestHandler handler, Class<? extends AppiumParams> paramClass) throws DuplicateRouteException {
         if (!routerMap.containsKey(method)) {
             routerMap.put(method, new HashMap<String, RequestHandler>());
         }
         if (routerMap.get(method).containsKey(uri)) {
             throw new DuplicateRouteException();
         }
+        if (!paramClassMap.containsKey(method)) {
+            paramClassMap.put(method, new HashMap<String, Class>());
+        }
 
         routerMap.get(method).put(uri, handler);
+        paramClassMap.get(method).put(uri, paramClass);
     }
 
-    BaseResponse route(IHTTPSession session) {
-        RequestHandler handler;
-        Map<String, String> uriParams;
+    AppiumResponse route(IHTTPSession session) {
+        RequestHandler handler = null;
 
-        try {
-            String uri = session.getUri();
-            Method method = session.getMethod();
+        String uri = session.getUri();
+        Method method = session.getMethod();
 
-            System.out.println("Received " + method + " request for '" + uri + "'");
+        System.out.println("Received " + method + " request for '" + uri + "'");
 
-            if (!routerMap.containsKey(method)) {
-                routerMap.put(method, new HashMap<String, RequestHandler>());
-            }
+        if (!routerMap.containsKey(method)) {
+            routerMap.put(method, new HashMap<String, RequestHandler>());
+        }
+        if (!paramClassMap.containsKey(method)) {
+            paramClassMap.put(method, new HashMap<String, Class>());
+        }
 
-            // By default, set handler to NotFound until we find a matching handler
-            handler = new RequestHandler() {
-                @Override
-                public BaseResponse handle(IHTTPSession session, Map<String, String> uriParams) {
-                    return new NotFoundResponse();
+        Map<String, String> uriParams = new HashMap<>();
+
+        // Look for a matching route
+        // TODO: Move this to a separate method 'isRouteMatch'.
+        Class paramClass = AppiumParams.class;
+        for (Map.Entry<String, RequestHandler> entry : routerMap.get(method).entrySet()) {
+            String testUri = entry.getKey();
+
+            // TODO: Use StringBuilder to construct the Test Regexes
+            String testRegex = "^";
+            Map<Integer, String> wildcardIndices = new HashMap<>();
+
+            // Convert route to a regex to test incoming URI against
+            // TODO: Cache these regexes instead of re-creating them every time
+            int index = 0;
+            for (String uriToken : testUri.split("/")) {
```

Comment:
this method is a bit hard to read. Can we extract parsing operations into separate private helper methods?